### PR TITLE
Update to v0.10.1 and add build for Apple Silicon

### PR DESCRIPTION
**Pull request by wietsedv**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

This is an update to v0.10.1 and added compilation for Apple Silicon.

[Edit by @Anthchirp: Closes #19]

### CODE REVIEW

**Comment by Anthchirp** [comment]
The recommended way to add the architecture support is to add the package to the file https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/migrations/osx_arm64.txt via a pull request, and then the bots will take care of the rest, including the upstream dependency tree.

(Apologies if you knew this and did it your way on purpose)

**Comment by wietsedv** [comment]
> The recommended way to add the architecture support is to add the package to the file https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/migrations/osx_arm64.txt via a pull request, and then the bots will take care of the rest, including the upstream dependency tree.
> 
> (Apologies if you knew this and did it your way on purpose)

Thanks! I did not know, and I will certainly use that method for any future migrations.

But in any case, this PR also upgrades to the latest tokenizers version (which the tickbot failed to do in #15). And the Rust-python bindings with PyO3 requires special attention. The full error is:

    error: failed to run custom build command for `pyo3 v0.12.4`
  
    Caused by:
      process didn't exit successfully: `/private/tmp/pip-req-build-gkfo3s_s/target/release/build/pyo3-14aa9f9df061ffef/build-script-build` (exit code: 1)
      --- stderr
      Error: "Must provide PYO3_CROSS_LIB_DIR environment variable when cross-compiling"
    warning: build failed, waiting for other jobs to finish...
    error: build failed
    error: cargo failed with code: 101

PyO3 expects `PYO3_CROSS_LIB_DIR` to be set to "the directory containing the target's libpython DSO": https://pyo3.rs/v0.12.4/building_and_distribution.html#cross-compiling

I can take a look at how to do this later, but feel free to help if anyone has actual experience with this.

**Comment by Anthchirp** [comment]
I see that you got the builds working for everything but osx_arm64. Is it worth splitting up the PR and merging the version update first?

**Comment by wietsedv** [comment]
> I see that you got the builds working for everything but osx_arm64. Is it worth splitting up the PR and merging the version update first?

I might have fixed it. Sorry for the many commits, but I can not debug this cross compilation issue locally.

**Comment by wietsedv** [comment]
Current problems:
 - Python3.9 does not build because of a version requirement problem I have not yet fully investigated.
 - Python3.8 does not build because PyO3 is confused.

PyO3 searches for `$PREFIX/lib/python3.*/_sysconfigdata_*.py` and it finds multiple matches. For some reason sysconfigdata files for other platforms are available and PyO3 does not like that.

How to get rid of files like these while building osx-arm64:
 - `$PREFIX/lib/python3.8/_sysconfigdata_i686_conda_linux_gnu.py`
 - `$PREFIX/lib/python3.8/_sysconfigdata_powerpc64le_conda_cos7_linux_gnu.py`
 - etc.

**Comment by wietsedv** [comment]
I now have a working osx-arm64 build with Python 3.8. This included adding the following workaround to the build script:

    find {{ PREFIX }}/lib/python3.* -name "_sysconfigdata_*.py" -type f -not -name "_sysconfigdata_arm64_*.py" -delete  # [arm64]

This is necessary because PyO3 (rust python bridge) searches for any `sysconfigdata` files in the lib directory and breaks if there are multiple. In normal circumstances there is only one variant that corresponds to the target platform, but Anaconda python includes each possible platform variant. Python internally uses the `_PYTHON_SYSCONFIGDATA_NAME` environment variable workaround for this particular issue, but I do not think this is currently solvable for PyO3.

My workaround only affects the throwaway build environment, so I think it is acceptable. If this is merged, I'll create an issue in the PyO3 repository and also an issue here that links to that PyO3 issue, so that this workaround will not be forgotten.

Additionally, I have disabled the build for Python 3.9 since there seems to be some version conflict where a dependency has not OSX_ARM64 Python 3.9 support yet. I can make a PR for 3.9 after this one is merged for 3.9 support (which I can rebase on master if the issue magically disappears.). In any case just having 3.8 support is better than having nothing.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -15,10 +15,18 @@ build:
...
     - {{ PYTHON }} -m pip install . -vv
+  skip: True  # [arm64 and py==39]
 
 requirements:
   build:
+    - python                                 # [build_platform != target_platform]
+    - cross-python_{{ target_platform }}     # [build_platform != target_platform]
+    - cython                                 # [build_platform != target_platform]
+    - numpy                                  # [build_platform != target_platform]
+    - pybind11                               # [build_platform != target_platform]
```
Comment (Anthchirp) [inline recipe/meta.yaml:29]: Do these dependencies need to be added to build? The ['official' migration](https://github.com/conda-forge/tokenizers-feedstock/pull/19) doesn't do this

**Review by Anthchirp — APPROVED** [review]

File: recipe/meta.yaml (hunk 1 above)
Comment (wietsedv) [reply recipe/meta.yaml]: I don't think it actually is. I natively copied from here: https://conda-forge.org/blog/posts/2020-10-29-macos-arm64/#how-to-add-a-osx-arm64-build-to-a-feedstock

Also, I think that #19 does not have the Python 3.9 issue, so let's see whether removing these lines resolves the issue.

**Comment by wietsedv** [comment]
Every build is succesfull now, including OSX_ARM64 Py39!

**Comment by wietsedv** [comment]
I created an PyO3 issue for resolving the `sysconfigdata` removal workaround: https://github.com/PyO3/pyo3/issues/1484

I guess it will take a while before a release is fixed, so I think that this PR is done. :)

**Comment by Anthchirp** [comment]
Thank you very much!

**Pull request merged**